File: packages/go-sdk/composition/trigger.go
package composition

import "context"

// TriggerAndWait triggers a run and polls until it reaches a terminal status.
func TriggerAndWait[TInput any, TRun any](
	ctx context.Context,
	triggerFn func(ctx context.Context, input TInput) (TRun, error),
	getRun func(ctx context.Context, runID string) (TRun, error),
	getID func(TRun) string,
	getStatus func(TRun) string,
	input TInput,
	opts *WaitForRunOptions,
) (TRun, error) {
	run, err := triggerFn(ctx, input)
	if err != nil {
		var zero TRun
		return zero, err
	}
	return WaitForRun(ctx, getRun, getStatus, getID(run), opts)
}

File: packages/rust-sdk/src/composition/trigger.rs
use super::wait::{WaitForRunOptions, wait_for_run};
use crate::errors::StraitError;
use std::future::Future;

pub async fn trigger_and_wait<
    TInput,
    TRun,
    TriggerFut,
    GetRunFut,
    TriggerFn,
    GetRunFn,
    GetIdFn,
    GetStatusFn,
>(
    input: TInput,
    trigger_fn: TriggerFn,
    get_run: GetRunFn,
    get_id: GetIdFn,
    get_status: GetStatusFn,
    opts: Option<&WaitForRunOptions>,
) -> Result<TRun, StraitError>
where
    TriggerFn: FnOnce(TInput) -> TriggerFut,
    TriggerFut: Future<Output = Result<TRun, StraitError>>,
    GetRunFn: Fn(&str) -> GetRunFut,
    GetRunFut: Future<Output = Result<TRun, StraitError>>,
    GetIdFn: FnOnce(&TRun) -> &str,
    GetStatusFn: Fn(&TRun) -> &str,
{
    let result = trigger_fn(input).await?;
    let run_id = get_id(&result).to_string();
    wait_for_run(&run_id, get_run, get_status, opts).await
}
